fv3gfs/util/halo_data_transformer.py

```diff
@@ -0,0 +1,905 @@
+        if __debug__:
+            if len(quantities_x) != len(self._infos_x) and len(quantities_y) != len(
+                self._infos_y
+            ):
+                raise RuntimeError(
+                    f"Quantities count (x: {len(quantities_x)}, y: {len(quantities_y)}"
+                    f" is different that specifications count (x: {len(self._infos_x)}, y: {len(self._infos_y)}"
+                )
+            # TODO Per quantity check
+        assert isinstance(self._pack_buffer, Buffer)  # e.g. allocate happened
+        assert len(self._infos_x) == len(self._infos_y)
+        assert len(quantities_x) == len(quantities_y)
+        offset = 0
+        for quantity_x, quantity_y, info_x, info_y, in zip(
+            quantities_x, quantities_y, self._infos_x, self._infos_y
+        ):
+            cu_kernel_args = self._cu_kernel_args[info_x._id]
+
+            # Use private stream
+            with self._get_stream(cu_kernel_args.stream):
+
+                # Buffer sizes
+                transformer_size = info_x._pack_buffer_size + info_y._pack_buffer_size
+
+                if quantity_x.metadata.dtype != np.float64:
+                    raise RuntimeError(f"Kernel requires f64 given {np.float64}")
+
+                # Launch kernel
+                blocks = 128
+                grid_x = (transformer_size // blocks) + 1
+                if pack_vector_f64_kernel is None:
+                    RuntimeError("CUDA nvrtc failed")
+                else:
+                    pack_vector_f64_kernel(
+                        (blocks,),
+                        (grid_x,),
+                        (
+                            quantity_x.data[:],  # source_array_x
+                            quantity_y.data[:],  # source_array_y
+                            cu_kernel_args.x_send_indices,  # indices_x
+                            cu_kernel_args.y_send_indices,  # indices_y
+                            info_x._pack_buffer_size,  # nIndex_x
+                            info_y._pack_buffer_size,  # nIndex_y
+                            offset,
+                            (-info_x.pack_clockwise_rotation) % 4,  # rotation
+                            self._pack_buffer.array,
+                        ),
+                    )
+
+                # Next transformer offset into send buffer
+                offset += transformer_size
+
+    def async_unpack(
+        self,
+        quantities_x: List[Quantity],
+        quantities_y: Optional[List[Quantity]] = None,
+    ):
+        """Unpack the quantities from a single buffer via streamed cuda kernels
+        
+        Reads from self._unpack_buffer using self._x_infos and self._y_infos
+        to read the offsets and sizes per quantity.
+
+        Args:
+            quantities_x: list of quantities to unpack. Must fit the specifications given
+            at init time.
+            quantities_y: Same as above but optional, used only for Vector transfer.
```

Comment:
Removed capital since vector is not a class name.
```suggestion
            quantities_y: Same as above but optional, used only for vector transfer.
```

Reply:
Done